Add Cassandra and ScyllaDB docs (#17833)
gravitational/teleport: docs/pages/database-access/guides/aws-cassandra-keyspaces.mdx

---

> title: Database Access with Cassandra and ScyllaDB

Reviewer: ```suggestion
title: Database Access with AWS Keyspaces (Apache Cassandra)
```

A few other places in this guide mention "Cassandra and ScyllaDB" instead of "AWS Keyspaces"). Should we change it to match the guide link/content?

<ScopedBlock scope={["oss", "enterprise"]}>
![Teleport Database Access Redis Self-Hosted](../../../img/database-access/guides/cassandra_keyspaces_slefhosted.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>
![Teleport Database Access Redis Cloud](../../../img/database-access/guides/cassandra_keyspaces_cloud.png)
</ScopedBlock>

## Prerequisites

- Self-hosted Cassandra or ScyllaDB instance.
- cqlsh cassandra client installed and added to your system's `PATH` environment variable.
- - A host, e.g., an Amazon EC2 instance, where you will run the Teleport Database
Service.

(!docs/pages/includes/user-client-prereqs.mdx!)

## Step 1/5. Install and configure Teleport

### Set up the Teleport Auth and Proxy Services

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

### Set up the Teleport Database Service

(!docs/pages/includes/database-access/token.mdx!)

Install Teleport on the host where you will run the Teleport Database Service:

(!docs/pages/includes/install-linux.mdx!)

<ScopedBlock scope={["oss", "enterprise"]}>

Start the Teleport Database Service, pointing the `--auth-server` flag to the
address of your Teleport Proxy Service:

```code

> $teleport db start \

Reviewer: Question: Is this the default manner of starting the database service? I initially thought it was the configuration file approach with `teleport db configure create ...` and `teleport start -c`.

> $teleport db start \

Author: I don't think that we have a defautl way. 
For self hosted database we recumbent to use `teleport start db` approach where as for AWS related db services  if auto discovery is supported `teleport db configure` should be preferred. The AWS Keystores is quite different because you can have only it enabled or disable so autodiscovery flow doesn't fit here.

  --token=/tmp/token \
  --auth-server=teleport.example.com:443 \
  --name=keyspaces \
  --protocol=cassandra \
  --aws-account-id=12345678912 \
  --aws-region=us-east-1 \
  --labels=env=dev
```

<Admonition type="note">

The `--auth-server` flag must point to the Teleport cluster's Proxy Service
endpoint because the Database Service always connects back to the cluster over a
reverse tunnel.

</Admonition>

</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

Start the Teleport Database Service, pointing the `--auth-server` flag to the
address of your Teleport Cloud tenant:

```code
$ teleport db start \
  --token=/tmp/token \
  --auth-server=mytenant.teleport.sh:443 \
  --name=keyspaces \
  --protocol=cassandra \
  --aws-account-id=12345678912 \
  --aws-region=us-east-1 \
  --labels=env=dev
```

</ScopedBlock>

<Admonition type="tip">
  You can start the Database Service using a configuration file instead of CLI flags.
  See the [YAML reference](../reference/configuration.mdx) for details.
</Admonition>

## Step 2/5. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/5. [Optional] Create AWS Keyspaces role:

> Go to the IAM -> Access Management -> [Roles](https://console.aws.amazon.com/iamv2/home#/roles).

Reviewer 2: nit: I feel this part can use a little more context, like what the role does, or maybe how teleport will use them, or who can skip them.

i assume users likely already have roles for their access. do we recommend them use existing roles or create new roles for teleport?

Author: Added a sentence  describing how teleport will use this role.  

> I assume users likely already have roles for their access. do we recommend them use existing roles or create new roles for teleport?

Not sure if that is always true.  In case of new Keyspaces setup a user will need to create a new AWS Keyspaces role.

Press Create Role.

> ![Create Role Step 1](../../../img/database-access/guides/keyspaces/create-role-step1.png)

Reviewer 2: hmm, the create role page looks a different for me
<img width="900" alt="Screenshot 2022-10-27 at 9 36 21 AM" src="https://user-images.githubusercontent.com/12417665/198311589-3effa666-7b08-44e5-ab8b-a2730d08f890.png">
Not a big deal, but i assume i am seeing a newer version which is slightly more user-friendly where you can select your own account.

> Create custom [IAM Keyspaces Policy](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html) or select already existed `AmazonKeyspacesReadOnlyAccess` policy.

Reviewer 2: ```suggestion
Create custom [IAM Keyspaces Policy](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html) or select an AWS-managed policy like `AmazonKeyspacesReadOnlyAccess`.
```

You can choose the `AmazonKeyspacesReadOnlyAccess` for the AWS Keyspaces read-only access or `AmazonKeyspacesFullAccess` for the AWS Keyspaces full access.
<Admonition type="tip">

>     The `AmazonKeyspacesReadOnlyAccess` and `AmazonKeyspacesReadOnlyAccess` policies may provide too much or not enough access for your intentions. Validate these meet your expectations if you plan on using them. You can also create your own custom AWS Keyspaces Permissions Policies: [Amazon Keyspaces identity-based policy examples](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html).

Reviewer 3: ```suggestion
The `AmazonKeyspacesReadOnlyAccess` and `AmazonKeyspacesReadOnlyAccess` policies may
provide too much or not enough access for your intentions. 
Validate that these meet your expectations if you plan on using them. 
You can also create your own custom AWS Keyspaces Permissions Policies: [Amazon Keyspaces identity-based policy examples](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html).
```
This suggestion adds a "that" after "Validate" and breaks up this paragraph for easier editing later on

</Admonition>
Confirm role
Enter a role name and press create role.
![Create Role Step 1](../../../img/database-access/guides/keyspaces/create-role-step3.png)

## Step 4/5. Give Teleport permissions to assume roles

Next, attach the following policy to the IAM role or IAM user the Teleport
Database Service instance is using, which allows the Database Service to
assume the IAM roles:

```yaml
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "sts:AssumeRole",
      "Resource": "*"
    }
  ]
}
```

<Admonition type="tip">
  You can make the policy more strict by providing specific IAM role resource
  ARNs in the "Resource" field instead of using a wildcard.
</Admonition>

## Step 5/5. Connect

Once the Database Service has joined the cluster, log in to see the available
databases:

<ScopedBlock scope={["oss", "enterprise"]}>

  ```code

>   $ tsh login --proxy=teleport.example.com --user=testuser

Reviewer 2: The `Create a Teleport user` step creates a user named `alice`
```suggestion
  $ tsh login --proxy=teleport.example.com --user=alice
```

  $ tsh db ls
  # Name      Description Allowed Users Labels  Connect
  # --------- ----------- ------------- ------- -------
  # keyspaces             [*]           env=dev
  ```

</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

  ```code
  $ tsh login --proxy=mytenant.teleport.sh --user=testuser
  $ tsh db ls
  # Name      Description Allowed Users Labels  Connect
  # --------- ----------- ------------- ------- -------
  # keyspaces             [*]           env=dev
  ```

</ScopedBlock>

To connect to a particular database instance :
```code

> $ tsh db connect --db-user=KeyspacesReader keyspaces

Reviewer 2: If not explained already in previous steps, we should note here the `KeyspacesReader` is the AWS role name.

# Connected to Amazon Keyspaces at localhost:55084
# [cqlsh 6.0.0 | Cassandra 3.11.2 | CQL spec 3.4.4 | Native protocol v4]
# Use HELP for help.
# KeyspacesReader@cqlsh>
```

To log out of the database and remove credentials:

```code
# Remove credentials for a particular database instance.

> $ tsh db logout example

Reviewer 2: ```suggestion
$ tsh db logout keyspaces
```

---
title: Database Access with AWS Keyspaces (Apache Cassandra)

> description: How to configure Teleport Database Access with Azure Database for AWS Keyspaces (Apache Cassandra)

Reviewer 3: ```suggestion
description: How to configure Teleport Database Access with AWS Keyspaces (Apache Cassandra)
```

---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="11.0"
>

>   Database access for AWS Keyspaces (Apache Cassandra) is available starting from Teleport `v11.0`.

Reviewer 3: ```suggestion
  Database Access for AWS Keyspaces (Apache Cassandra) is available starting from Teleport `v11.0`.
```

</Details>

This guide will help you to:

- Install and configure Teleport.

> - Set up Teleport to access your AWS Keyspaces (Apache Cassandra)..

Reviewer 3: ```suggestion
- Set up Teleport to access AWS Keyspaces (Apache Cassandra).
```

- Connect to your database through Teleport.

<ScopedBlock scope={["oss", "enterprise"]}>
![Teleport Database Access Redis Self-Hosted](../../../img/database-access/guides/cassandra_keyspaces_slefhosted.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>
![Teleport Database Access Redis Cloud](../../../img/database-access/guides/cassandra_keyspaces_cloud.png)
</ScopedBlock>

## Prerequisites

- AWS Account with AWS Keyspaces database and permissions to create and attach IAM policies

> - cqlsh Cassandra client installed and added to your system's `PATH` environment variable.

Reviewer 3: ```suggestion
- The `cqlsh` Cassandra client installed and added to your system's `PATH` environment variable.
```

address of your Teleport Proxy Service:

```code
$ teleport db start \
  --token=/tmp/token \
  --auth-server=teleport.example.com:443 \
  --name=keyspaces \
  --protocol=cassandra \
  --aws-account-id=12345678912 \
  --aws-region=us-east-1 \
  --labels=env=dev
```

<Admonition type="note">

The `--auth-server` flag must point to the Teleport cluster's Proxy Service
endpoint because the Database Service always connects back to the cluster over a
reverse tunnel.

</Admonition>

</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

Start the Teleport Database Service, pointing the `--auth-server` flag to the
address of your Teleport Cloud tenant:

```code
$ teleport db start \
  --token=/tmp/token \
  --auth-server=mytenant.teleport.sh:443 \
  --name=keyspaces \
  --protocol=cassandra \
  --aws-account-id=12345678912 \
  --aws-region=us-east-1 \
  --labels=env=dev
```

</ScopedBlock>

<Admonition type="tip">
  You can start the Database Service using a configuration file instead of CLI flags.
  See the [YAML reference](../reference/configuration.mdx) for details.
</Admonition>

## Step 2/5. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

> ## Step 3/5. [Optional] Create AWS Keyspaces role:

Reviewer 3: ```suggestion
## Step 3/5. [Optional] Create an AWS Keyspaces role
```

> Create an AWS IAM Role that will be used as Keyspaces user. If you already have a custom IAM Keyspaces role, you can skip this step.

Reviewer 3: ```suggestion
Create an AWS IAM Role that will be used as your Keyspaces user. If you already have a custom IAM Keyspaces role, you can skip this step.
```

> Create an AWS IAM Role that will be used as Keyspaces user. If you already have a custom IAM Keyspaces role, you can skip this step.

Reviewer 3: I think we should spell out the permissions we expect the role to have here. Otherwise, a user may use an existing role with overly broad or narrow permissions.

Author: I think that we can remove the part about reusing existing policy because it a bit misleading  and recommend to create a teleport access role based on `AmazonKeyspacesReadOnlyAccess` or `AmazonKeyspacesFullAccess` policies.

Go to the IAM -> Access Management -> [Roles](https://console.aws.amazon.com/iamv2/home#/roles).
Press Create Role.

![Create Role Step 1](../../../img/database-access/guides/keyspaces/create-role-step1.png)

> Create custom [IAM Keyspaces Policy](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html) or select an AWS-managed policy like `AmazonKeyspacesReadOnlyAccess` policy.

Reviewer 3: ```suggestion
Create a custom [IAM Keyspaces Policy](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html) or select an AWS-managed policy like `AmazonKeyspacesReadOnlyAccess`.
```

Reviewer 3: Should we make it more explicit that we recommend using the `AmazonKeyspacesReadOnlyAccess` policy, or which policies we recommend for different levels of access (e.g., read access or write access)? Otherwise, "an AWS-managed policy like" may be a bit too vague.

Author: Changed to 
```
Select `AmazonKeyspacesReadOnlyAccess` for read-only AWS Keyspaces access or `AmazonKeyspacesFullAccess` for full AWS Keyspaces access.
<Admonition type="tip">
    You can make the policy more strict by providing specific IAM role Cassandra action and specifying the Cassandra Keyspaces table [Amazon Keyspaces identity-based policy examples](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html).
</Admonition>
```

![Create Role Step 1](../../../img/database-access/guides/keyspaces/create-role-step2.png)
<Admonition type="tip">

>   You can make the policy more strict by providing specific IAM role cassandra action and specifying the cassandra keyspace and table [Amazon Keyspaces identity-based policy examples](https://docs.aws.amazon.com/keyspaces/latest/devguide/security_iam_id-based-policy-examples.html).

Reviewer 3: - We should capitalize "Cassandra"
- Is there another way to phrase, "providing specific IAM role cassandra action"? This seems a bit garbled to me.
- This tip should use complete sentences.

Author: updated

</Admonition>

> Confirm role

Reviewer 3: ```suggestion
```

> Enter a role name and press create role.

Reviewer 3: ```suggestion
Enter a role name and press "Create role".
```
